mne_sandbox/preprocessing/dss.py

```diff
@@ -0,0 +1,140 @@
+        bias_max_components=None, bias_thresh=0, return_data=True):
+    """Process physiological data with denoising source separation (DSS)
+
+    Implementation follows the procedure described in Särelä & Valpola [1]_
+    and de Cheveigné & Simon [2]_.
+
+    Parameters
+    ----------
+    data : instance of Epochs | array of shape (n_trials, n_channels, n_times)
+        Data to be denoised.
+    data_max_components : int | None
+        Maximum number of components to keep during PCA decomposition of the
+        data. ``None`` (the default) keeps all suprathreshold components.
+    data_thresh : float | None
+        Threshold (relative to the largest component) above which components
+        will be kept during decomposition of the data. The default keeps all
+        non-zero values; to keep all values, specify ``thresh=None``.
+    bias_max_components : int | None
+        Maximum number of components to keep during PCA decomposition of the
+        bias function. ``None`` (the default) keeps all suprathreshold
+        components.
+    bias_thresh : float | None
+        Threshold (relative to the largest component) below which components
+        will be discarded during decomposition of the bias function. ``None``
+        (the default) keeps all non-zero values; to keep all values, pass
+        ``thresh=None`` and ``max_components=None``.
+    return_data : bool
+        Whether to return the denoised data along with the denoising matrix.
+
+    Returns
+    -------
+    dss_mat : array of shape (n_channels, n_dss_channels)
+        The denoising matrix.
+    dss_data : array of shape (n_trials, n_dss_channels, n_times)
+        The denoised data. Note that n_dss_channels are orthogonal virtual
+        channels and may be fewer in number than the number of channels in the
+        input Epochs object. Returned only if ``return_data`` is ``True``.
+
+    References
+    ----------
+    .. [1] Särelä, Jaakko, and Valpola, Harri (2005). Denoising source
+    separation. Journal of Machine Learning Research 6: 233–72.
+
+    .. [2] de Cheveigné, Alain, and Simon, Jonathan Z. (2008). Denoising based
+    on spatial filtering. Journal of Neuroscience Methods, 171(2): 331-339.
+    """
+    if isinstance(data, Epochs):
+        data_cov = compute_covariance(data).data
+        bias_cov = np.cov(data.average().data)
+        if return_data:
+            data = data.get_data()
+    elif isinstance(data, np.ndarray):
+        data_cov = np.sum([np.dot(trial, trial.T) for trial in data], axis=0)
+        bias_cov = np.cov(data.mean(axis=0))
+    else:
+        raise TypeError('Data to denoise must be an instance of mne.Epochs or '
+                        'a numpy array.')
+    dss_mat = _dss(data_cov, bias_cov, data_max_components, data_thresh,
+                   bias_max_components, bias_thresh)
+    if return_data:
+        n_trials, n_chans, n_times = data.shape
+        dss_data = np.einsum('ij,hik->hjk', dss_mat, data)
```

Comment:
people usually get confused with `einsum` so it might be worth adding a comment it's equivalent to `[np.dot(dss_mat, ep) for ep in data]`, even though it can be deduced from the indices